packages/react-instantsearch-nextjs/src/useInstantSearchRouting.ts

```diff
@@ -21,55 +20,51 @@ export function useInstantSearchRouting<
   const routingRef =
     useRef<InstantSearchProps<TUiState, TRouteState>['routing']>();
   const onUpdateRef = useRef<() => void>();
+  
+  useEffect(() => {
+    if (passedRouting && !routingRef.current) {
+      let browserHistoryOptions: Partial<BrowserHistoryArgs<TRouteState>> = {};
+  
+      browserHistoryOptions.getLocation = () => {
+        if (isMounting.current) {
+          return new URL(
+            `${window.location.protocol}//${window.location.host}${pathname}?${searchParams}`
+          ) as unknown as Location;
+        }
+  
+        return window.location;
+      };
+      browserHistoryOptions.push = function push(
+        this: ReturnType<typeof historyRouter>,
+        url
+      ) {
+        // This is to skip the push with empty routeState on dispose as it would clear params set on a <Link>
+        if (this.isDisposed) {
+          return;
+        }
+        router.push(url, { scroll: false });
+      };
+      browserHistoryOptions.start = function start(onUpdate) {
+        onUpdateRef.current = onUpdate;
+      };
+  
+      routingRef.current = {};
+      if (typeof passedRouting === 'object') {
+        browserHistoryOptions = {
+          ...browserHistoryOptions,
+          ...passedRouting.router,
+        };
+        routingRef.current.stateMapping = passedRouting.stateMapping;
+      }
+      routingRef.current.router = historyRouter(browserHistoryOptions);
+    }
+  }, []);
+  
   useEffect(() => {
     if (onUpdateRef.current) {
       onUpdateRef.current();
     }
   }, [pathname, searchParams]);
 
-  if (passedRouting && !routingRef.current) {
-    let browserHistoryOptions: Partial<BrowserHistoryArgs<TRouteState>> = {};
-
-    browserHistoryOptions.getLocation = () => {
-      if (typeof window === 'undefined') {
-        const url = `${
-          headers().get('x-forwarded-proto') || 'http'
-        }://${headers().get('host')}${pathname}?${searchParams}`;
-        return new URL(url) as unknown as Location;
-      }
```

Comment:
since this is a client component, we can't call headers in here for app router. Also, this will be unnecessary for a client component to access `window` context